### DISCUSSION TITLE
Include Terraform Plugins in build

### DISCUSSION THREAD

**Discussion by tamirkamara** [getporter/porter#2279] [Q&A]
Hi,

Do you have some guidance on how one can include terraform plugins in the bundle (during build time)?

I found [this doc](https://www.terraform.io/cli/commands/providers/mirror) from Terraform and it might work if include in the bundle's docker file. 
However, I see an issue since `terraform providers mirror` requires an init beforehand to discover local modules. Should I also run terraform init in the bundle's docker file, or do you run it automatically?

Any guidance would be appreciated.

**Comment by tamirkamara** [comment]
I think I can run the following in the bundle dockerfile, but is that the right thing todo given the mixin also runs terraform init?
`
RUN cd $BUNDLE_DIR/terraform && terraform init -backend=false && terraform providers mirror /usr/local/share/terraform/plugins
`

**Comment by carolynvs** [comment]
Disclaimer: I don't use terraform 😀 

From Terraform's docs for `terraform init`

> When terraform init is run, Terraform reads configuration files in the working directory to determine which plugins are necessary, searches for installed plugins in several locations, sometimes downloads additional plugins, decides which plugin versions to use, and writes a lock file to ensure Terraform will use the same plugin versions in this directory until terraform init runs again.

Based on feedback from other people using terraform in bundles, I have always thought that was sufficient and that you don't need to do extra commands to ensure that the bundle has all the necessary files locally in order to execute your terraform plans.

When you run your bundle using just the default behavior of the terraform mixin (which calls terraform init in the [configured directory and optionally using a custom providers file](https://github.com/getporter/terraform-mixin#initfile)), what is the error that you are seeing that prompted you to look into that other terraform providers mirror command?

**Reply by tamirkamara** [reply]
When calling install on the bundle terraform tries to connect to hashicorp domains. We had to add firewall rules for those to be allowed.
I wonder... If init should be enough then why does terraform have a separate mirror command?

**Reply by carolynvs** [reply]
Are you using a remote backend?

**Reply by tamirkamara** [reply]
I use the azurerm backend. I think is a remote one...

I think I have found part of the issue I'm seeing - it's the `initFile` settings. When that is specified porter only copies that file and then runs `terraform init` which is enough in simple situations but not when modules are used. The example below shows the difference in "discovered" providers. 
We have each module specify its own provider requirements and the root one doesn't aggregate all the child ones. This means that the providers.tf file in the root module (terraform folder) doesn't specify all module requirements. I believe this practice is right per [this terraform doc](https://www.terraform.io/language/modules/develop/providers).
For now, I think I'll remove the initFile setting from our porter.yaml and retest things.

With initFile:
```
Step 4/16 : COPY terraform/providers.tf $BUNDLE_DIR/terraform/
 ---> eb7b99fe0cd8
Step 5/16 : RUN cd $BUNDLE_DIR/terraform && terraform init -backend=false
 ---> Running in 2fc0d6eb10f3

Initializing provider plugins...
- Finding hashicorp/null versions matching "3.1.1"...
- Finding hashicorp/azurerm versions matching "3.12.0"...
- Finding hashicorp/azuread versions matching "2.20.0"...
- Installing hashicorp/null v3.1.1...
- Installed hashicorp/null v3.1.1 (signed by HashiCorp)
- Installing hashicorp/azurerm v3.12.0...
- Installed hashicorp/azurerm v3.12.0 (signed by HashiCorp)
- Installing hashicorp/azuread v2.20.0...
- Installed hashicorp/azuread v2.20.0 (signed by HashiCorp
```

Without initFile:
```
Step 8/16 : COPY terraform/ $BUNDLE_DIR/terraform/
 ---> 1a7197b56304
Step 9/16 : RUN cd $BUNDLE_DIR/terraform && terraform init -backend=false
 ---> Running in 5a5e74d8b206
Initializing modules...
- aad in aad
- airlock in airlock
- azure_monitor in azure-monitor
- network in network

Initializing provider plugins...
- Reusing previous version of hashicorp/random from the dependency lock file
- Reusing previous version of hashicorp/azurerm from the dependency lock file
- Reusing previous version of hashicorp/azuread from the dependency lock file
- Reusing previous version of hashicorp/null from the dependency lock file
- Reusing previous version of hashicorp/local from the dependency lock file
- Installing hashicorp/random v3.3.2...
- Installed hashicorp/random v3.3.2 (signed by HashiCorp)
- Installing hashicorp/azurerm v3.12.0...
- Installed hashicorp/azurerm v3.12.0 (signed by HashiCorp)
- Installing hashicorp/azuread v2.20.0...
- Installed hashicorp/azuread v2.20.0 (signed by HashiCorp)
- Installing hashicorp/null v3.1.1...
- Installed hashicorp/null v3.1.1 (signed by HashiCorp)
- Installing hashicorp/local v2.2.3...
- Installed hashicorp/local v2.2.3 (signed by HashiCorp)
```

**Reply by tamirkamara** [reply]
@carolynvs so my theory didn't work :-( still seeing the error.
```
Error: could not init terraform, exit status 1
Error: mixin execution failed: exit status 1
Error: 2 errors occurred:
* container exit code: 1, message: <nil>. fetching outputs failed: error copying outputs from container: Error: No such container:path: d776b1d1ee4a935d0c77922d1fd4fdbfecdd7dcc996b97795fe1a2ade260a1f2:/cnab/app/outputs
* required output app_role_id_workspace_airlock_manager is missing and has no default
```

Is there another `init` while executing install?

**Reply by tamirkamara** [reply]
So it does look like it runs init again. Possible this is it?

```
Initializing Terraform...
/usr/bin/terraform terraform init -backend=true -backend-config=container_name=tfstate -backend-config=key=42731aab-f352-40e0-a726-4fc02ef7b59f -backend-config=resource_group_name=rg-foo -backend-config=storage_account_name=foo -reconfigure
Initializing modules...

Initializing the backend...
```

**Reply by carolynvs** [reply]
Thanks for sticking with me, and reporting this problem! Again I don't use terraform so I'm trying to learn as I go! 😅 

Yup, that's why I asked about the remote backend. Here's the issue where we have been collecting info on how to better handle that remote backends require that terraform init is called when the bundle runs.

https://github.com/getporter/terraform-mixin/issues/80

I wasn't aware of the mirror functionality but after reading a few docs on it and the old [bundle](https://github.com/hashicorp/terraform/blob/main/tools/terraform-bundle/README.md) functionality, I understand now that is Hashicorp's official solution to this problem.

Long term, I think(?) that the terraform mixin should call mirror by default so that we don't re-download everything when it runs. Short term, yes that command that you are running seems like the right workaround. I believe the [custom dockerfile](https://getporter.org/bundle/custom-dockerfile/) should give you everything you need to do that, and you can move around the special comments like `# PORTER_MIXINS` if you need to change the order of when certain commands are run. If not, let me know and I promise I am more helpful about how to make porter do stuff than how terraform works! 🤦‍♀️ 

If you can help me figure out which command that the mixin should be calling by default when we build the bundle, or if we need additional configuration, let me know and we can get the terraform mixin updated to do the right thing!

**Reply by tamirkamara** [reply]
Thanks @carolynvs!
I did a couple of tests and looks like the mirror one works just fine. We have a large solution which is deployed multiple times each day (part of cicd), so we might be able to give this a try for a while and see how it goes.

As a side note, I will offer a PR on the mixin readme to note that the initFile property should be used carefully when having submodules.

**Reply by carolynvs** [reply]
I created https://github.com/getporter/terraform-mixin/issues/90 to track adding the mirror command to the default commands run by the mixin during porter build.
